bench/bm_test_helper.hpp

```diff
@@ -22,6 +22,19 @@ class source : public fg::node<source<T, min, count>> {
 
     source(std::size_t n_samples) : _n_samples_max(n_samples) {}
 
+    template<typename Self>
+    [[nodiscard]] constexpr auto
+    inputs_status(Self &self) noexcept {
+        const auto avail = _n_samples_max - n_samples_produced;
+
+        const struct {
```

Comment:
I'd propose naming this struct and moving it to node.hpp - easier to understand for the user 'Source nodes must implement input_status_t input_status(...)' than what is currently written in the static_assert.

Reply:
Needed to think about this a bit more but I concur with @ivan-cukic.

The user-facing API should be as simple and as crisp as possible.

Outside the benchmark/unit tests merging of source-type node is rare and probably never happens.
This makes the comparison with the more common 'runtime' use case difficult. 
With the exception of benchmark/unit tests, most of the sources also do not know on start-up how many samples they are going to produce this can be anything from 0 to infinity.

If the structure is needed/helpful to minimise internal boiler-plate code, I'd suggest making it a node internal implementation that expert users may override if necessary (1-2% use-case IMO).

Reply:
Yes, I was too quick not to name this struct. Will do.

I believe it is wrong for all sources to (almost) unconditionally, always say "I can give you only one sample, no more, *no less* even".

I also believe that merging sources with a gain, normalize, etc. doesn't sound too far fetched, no?

Wrt. sources being able to answer the question at all: I think they should be able to say "well I have 1K in my hardware input buffer ready to read, so ... 256 samples ready to go". They don't have to say how many samples they will produce in their lifetime. It's just that 32 can be optimized so much better than 1.

If all this doesn't convince you, then there should not be any benchmarks that merge the source node. Because if they exist, you're comparing apples with oranges wrt. SIMD. That's misleading at best.

I believe the `inputs_status` could be simplified for nodes that don't need as much control (especially nodes with a single output port). These could simply return a single `size_t`. What about an `available_samples` function that is only called if the node has a single output port? The could also be a `static_assert` to ensure no multi-output-port node implements the function.

Reply:
Trying to understand the implications wrt. `graph::work`: Do I understand correctly, that, without this change, a source node's `work` function is called once *per single sample*... through an indirect function call (virtual function)? And since `graph::work` cannot take shortcuts in its loop it calls `work` on *all* nodes once per sample? (Unless, the source implements `work`, `process_bulk`, or `inputs_status`.)

Reply:
Hmm, even `process_bulk` shouldn't help because the generic `node::work` function only requests a single sample from output port writers?

Reply:
> Wrt. sources being able to answer the question at all: I think they should be able to say "well I have 1K in my hardware input buffer ready to read, so ... 256 samples ready to go". They don't have to say how many samples they will produce in their lifetime. It's just that 32 can be optimized so much better than 1.

For the dynamic/runtime case (inside `work()` this is queried and checked by the `size_t N = buffer.[reader, writer()].available()` expression.

If this is like `in case source block produces data, you can expect at least N samples' then there is the `std::size_t MIN_SAMPLES = std::dynamic_extent` NTTP on the port that maybe could be reused for that. 

We shouldn't forget that the merge-API is already helpful because it avoids writing intermediate results to a runtime buffer.

Reply:
I still don't understand. `graph::work` calls `node_wrapper<T>::work` which calls `T::work`. The latter is `node<Derived>::work` unless `Derived` implements `work`. If the source does *not implement* `work`, then the generic `node::work` implementation queries `inputs_status` which unconditionally returns `{ .at_least_one_input_has_data = false, .available_values_count = 1 }` for source nodes. The `at_least_one_input_has_data` value is ignored because `available_values_count` is non-zero. `work` then calls `output_port.writer().get(1)` for each output port. If the source node implements `process_bulk`, then `process_bulk(writers...)` is called and `available_values_count` are published to the output ports. Otherwise, `work` loops via `for (i = 0; i < 1; ++i)` over `process_one` calls and writes the result to the ports.
So, unless source nodes implement `work`, they produce exactly one sample with each iteration in `graph::work`.

Reply:
To summarize my understanding:
* A source node must either implement `work` or `inputs_status`.
* If a source node is to be merged at compile time it must implement `inputs_status`.

Since `inputs_status` is a very misleading function to call on source nodes, how about we introduce a different function? E.g. `size_t available_samples()`? However, how can the generic `node::work` implementation distinguish `work_return_t::INSUFFICIENT_INPUT_ITEMS` vs. `work_return_t::DONE`? `available_samples() == 0` could mean either one (where `INSUFFICIENT_INPUT_ITEMS` is a bad name again).